### ISSUE TITLE
 micro recall value correctness/explanation

### ISSUE THREAD

**Issue by kliegr** [AKSW/gerbil#88]
The micro recall statistics reported by Gerbil 1.1.1 using the A2KB type with "WEAK_ANNOTATION_MATCH" do not seem correspond to formulas (1), (3) in the BAT framework paper [1], referenced from the Metrics section of the Gerbil paper [2].

 To better trace the problem, I selected four documents with various tp/fp/fn statistics:
- Document 1 with one disambiguated entity in the groundtruth and none found by our system.
- Document 2 with one disambiguated entity in the groundtruth and the same entity detected and correctly disambiguated by our system.
- Document 3 with no entity in groundtruth and none assigned by our system.
- Document 4 with one disambiguated entity in the groundtruth, and one found by our system. Our system's annotation has the same beginIndex and endIndex as the groundtruth, but different disambiguation (different value of taIdentRef).

The micro recall returned by Gerbil is 0.5, although our system correctly disambiguated only one entity from the three present in the groundtruth across the four documents. This would suggest 0.33 micro recall. To double check, I recreated the documents along with the annotations in the GATE 8.0 Developer framework. The GATE Corpus Quality Assurance tool  also gives 0.33 recall.

To debug the problem, we performed Gerbil evaluation individually on each of the four documents. 
Of interest might be the  standalone A2KB Gerbil experiment on Document 4. Gerbil assigns Micro Recall "1.0" and Micro Precision "0.0", which is, by itself, a strange combination. GATE QA returns for this document zero precision and recall. This suggests that Gerbil can be counting one annotation as correct for recall and as incorrect for precision computation purposes.

I can send the sample documents and screenshots by email, I have them ready, but GitHub refuses to attach  them.

[1] http://static.googleusercontent.com/media/research.google.com/cs//pubs/archive/40749.pdf
[2] http://gerbil.aksw.org/gerbil/experiment?id=201411100001

**Comment by marcocor** [comment]
Hello,
My two cents: it seems you are right, micro-P and micro-R should be 1/2 and 1/3 respectively. This makes a micro-F1 of 1/(3*(1/3+1/2)) = 0.4 (instead of 0.5).

In your example, per-document measures should be (TP/FP/FN/P/R/F1):
Doc. 1: 0/0/1/1/0/0
Doc. 2: 1/0/0/1/1/1
Doc. 3: 0/0/0/1/1/1
Doc. 4: 0/1/1/0/0/0

mAcro-F1 is 0.5. Are you sure you are using mIcro rather than mAcro measures?

(About the corner-case you describe in the last paragraph about Doc 4: when the result returned by the system is empty, P is always 1 (This makes sense: there are no errors in the result). If the ground truth is empty as well, this makes R=1 and F1=1 (Which is consistent: the result is equal to the ground truth). If the ground truth is non-empty, R=0 and F1=0. So P and R for doument 4 are alright.)

Cheers,
Marco

**Comment by kliegr** [comment]
>When the result returned by the system is empty, P is always 1.  
For Document 4, the result is not empty. Both system and groundtruth have one entity at the same offsets, however the links on the entity are not matching.

**Comment by marcocor** [comment]
Sorry about that -- you are right again. Document 4 should have P=R=F1=0. There is definitely a bug in GERBIL.

**Comment by kliegr** [comment]
Thanks for very responsive feedback! Did you confirm against the NIF documents I tried to send you via email or via a separate check? To rule out other possible causes, such as problemes with parsing input document, etc.

**Comment by MichaelRoeder** [comment]
Hi all,

I am creating a JUnit test for the test case reported at the beginning of this issue. I think, I should be able to give you an update after lunch (in 1.5 hours).

**Comment by MichaelRoeder** [comment]
Could you please send me the NIF documents to roeder@informatik.uni-leipzig.de

**Comment by MichaelRoeder** [comment]
I added a simple class that runs the test described in the first post and the calculated values are correct:
```
Setting result of task 0 to ExperimentTaskResult(state=0,micF1=0.4000000059604645,micPrecision=0.5,micRecall=0.3333333432674408,macF1=0.5,macPrecision=0.75,macRecall=0.5,errors=0)
```

The problem seems to be the parsing of NIF documents. Our parser is not made to handle multiple `itsrdf:taIdentRef` properties for a single node. We have to fix that ;)
Together with the changes we are currently implementing for the OKE-Challenge in a branch of the project, your example should work as soon as we have finished these changes.

**Comment by kliegr** [comment]
Thanks for investigating this. Does this mean that the problem will be resolved if we correct the URI scheme and rerun the experiments?

**Comment by MichaelRoeder** [comment]
No, sorry.
It is a problem that occurs during the reading of the dataset and can't be solved that easely.

**Comment by kliegr** [comment]
Has there been some progress on this issue. If not, what is the ETA if it is on the roadmap?

**Comment by MichaelRoeder** [comment]
We are still working on this. Our current working plan contains a new version for the middle of may.

**Comment by kliegr** [comment]
Thanks for the update, looking forward to the new release!

**Comment by MichaelRoeder** [comment]
The problem of parsing multiple `itsrdf:taIdentRef` properties for a single annotation is solved in 1.2.0.
